src/EditorFeatures/Core/Implementation/KeywordHighlighting/HighlighterViewTaggerProvider.cs

```diff
@@ -3,74 +3,100 @@
     [ContentType(ContentTypeNames.CSharpContentType)]
     [ContentType(ContentTypeNames.VisualBasicContentType)]
     [TextViewRole(PredefinedTextViewRoles.Interactive)]
-    internal class HighlighterViewTaggerProvider :
-        AbstractAsynchronousViewTaggerProvider<HighlightTag>
+    internal class HighlighterViewTaggerProvider : AsynchronousTaggerProvider<KeywordHighlightTag>
     {
-        private readonly IHighlightingService _highlighterService;
+        private readonly IHighlightingService _highlightingService;
+
+        // Whenever an edit happens, clear all highlights.  When moving the caret, preserve 
+        // highlights if the caret stays within an existing tag.
+        public override TaggerCaretChangeBehavior CaretChangeBehavior => TaggerCaretChangeBehavior.RemoveAllTagsOnCaretMoveOutsideOfTag;
+        public override TaggerTextChangeBehavior TextChangeBehavior => TaggerTextChangeBehavior.RemoveAllTags;
+        public override IEnumerable<Option<bool>> Options => SpecializedCollections.SingletonEnumerable(InternalFeatureOnOffOptions.KeywordHighlight);
 
         [ImportingConstructor]
         public HighlighterViewTaggerProvider(
+            IHighlightingService highlightingService,
             IForegroundNotificationService notificationService,
-            IHighlightingService highlighterService,
             [ImportMany] IEnumerable<Lazy<IAsynchronousOperationListener, FeatureMetadata>> asyncListeners)
             : base(new AggregateAsynchronousOperationListener(asyncListeners, FeatureAttribute.KeywordHighlighting), notificationService)
         {
-            _highlighterService = highlighterService;
+            _highlightingService = highlightingService;
         }
 
-        protected override bool RemoveTagsThatIntersectEdits => true;
-
-        protected override SpanTrackingMode SpanTrackingMode => SpanTrackingMode.EdgeExclusive;
-
-        protected override IEnumerable<Option<bool>> TagSourceOptions
-        {
-            get
-            {
-                yield return InternalFeatureOnOffOptions.KeywordHighlight;
-            }
-        }
-
-        protected override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
+        public override ITaggerEventSource CreateEventSource(ITextView textView, ITextBuffer subjectBuffer)
         {
             return TaggerEventSources.Compose(
-                TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.OnIdle, reportChangedSpans: true),
+                TaggerEventSources.OnTextChanged(subjectBuffer, TaggerDelay.OnIdle),
                 TaggerEventSources.OnCaretPositionChanged(textView, subjectBuffer, TaggerDelay.NearImmediate),
                 TaggerEventSources.OnOptionChanged(subjectBuffer, FeatureOnOffOptions.KeywordHighlighting, TaggerDelay.NearImmediate));
         }
 
-        protected override ITagProducer<HighlightTag> CreateTagProducer()
+        // Internal for testing purposes
```

Comment:
Random thought that has this line get picked on by it: now since there's a ProduceTagsAsync function that's owned by the provider, can we have some central "give me tags I'm unit testing" function rather than having each one have to override this and explain why? Even just ProduceTagsAsync_ForUnitTestingOnly in the parent (still internal) might at least enforce the pattern versus require it to be replicated.